modules/asset-manager-storage-aws/src/main/java/org/opencastproject/assetmanager/aws/s3/AwsS3AssetStore.java

```diff
@@ -303,6 +447,90 @@ protected InputStream getObject(AwsAssetMapping map) throws AssetStoreException
+      if (expirationTime != null) {
+        return format("RESTORED, expires in %s", expirationTime.toString());
+      }
+
+      Boolean prevOngoingRestore = s3.getObjectMetadata(bucketName, map.getObjectKey()).getOngoingRestore();
+      if (prevOngoingRestore != null && prevOngoingRestore) {
+        return "RESTORING";
+      }
+
+      return "NONE";
+    } catch (AwsAssetDatabaseException | IllegalArgumentException e) {
+      throw new AssetStoreException(e);
+    }
+  }
+
+  /*
+   * Restore a frozen asset from deep archive
+   * @param storagePath asset storage path
+   * @param assetRestorePeriod number of days to restore assest for
+   * @see https://aws.amazon.com/s3/storage-classes/
+   */
+  public void initiateRestoreAsset(StoragePath storagePath, Integer assetRestorePeriod) throws AssetStoreException {
+    try {
+      AwsAssetMapping map = database.findMapping(storagePath);
+      restoreGlacierObject(map.getObjectKey(), assetRestorePeriod, false);
+    } catch (AwsAssetDatabaseException | IllegalArgumentException e) {
+      throw new AssetStoreException(e);
+    }
+  }
+
+  private void restoreGlacierObject(String objectName, Integer objectRestorePeriod, Boolean wait) {
+    Boolean prevOngoingRestore = s3.getObjectMetadata(bucketName, objectName).getOngoingRestore();
+    //FIXME: prevOngoingRestore is null when the object isn't being restored for some reason
+    // The javadocs for getOngoingRestore don't say anything about retuning null, and it doesn't make a ton of sense
+    // so I'm guessing this is a bug in the library itself that's not present in the version Manchester is using
+    if (prevOngoingRestore != null) {
+      if (prevOngoingRestore && wait) {
+        logger.info("Object {} is already being restored, waiting", objectName);
+      } else if (prevOngoingRestore && !wait) {
+        logger.info("Object {} is already being restored", objectName);
+        return;
+      }
+    }
+    RestoreObjectRequest requestRestore = new RestoreObjectRequest(bucketName, objectName, objectRestorePeriod);
+    s3.restoreObjectV2(requestRestore);
+
+    // if the object had already been restored the restore request will just
+    // increase the expiration time
+    if (s3.getObjectMetadata(bucketName, objectName).getRestoreExpirationTime() == null) {
+      logger.info("Restoring object {} from Glacier class storage", objectName);
+
+      // Just initiate restore?
+      if (!wait) {
+        return;
+      }
+
+      // Check the restoration status of the object.
+      // Wait min restore time and then poll ofter that
+      try {
+        // Check as restore might have already been initiated
+        if (prevOngoingRestore != null && !prevOngoingRestore) {
+          Thread.sleep(RESTORE_MIN_WAIT);
+        }
+
+        while (s3.getObjectMetadata(bucketName, objectName).getOngoingRestore()) {
+          Thread.sleep(RESTORE_POLL);
+        }
```

Comment:
This can potentially wait for hours, right? So an AssetManager getAsset() could time out and throw an exception, maybe even fail a workflow... Is there a better way to handle this? Maybe make use of the "availability" flag in the snapshot or something like that. I know this is a lot more work... :)

Reply:
Yes, absolutely will wait for hours.  Will take a look at the availability flag, but I'm not sure how useful that would be - the things trying to use the file(s) have to know about them too probably.  This may be a case of requiring business/institution logic (an admin rule that says 'don't do that') rather than properly supporting this in Opencast.  At least until the new new admin UI hits and someone funds the UI work to make this visible...

Reply:
Yes, you are correct. the `wait=true` option is not helpful if initiated by a http request, but this is meant to be an internal switch feature for code that knows what it's doing and is false when using the endpoints. I'm ok with it being removed. It is expected the  REST endpoints in this PR are to used manually ie someone initiates a restore and then checks whether it is available before initiating a workflow on the mediapackage. See my explanation here: https://github.com/opencast/opencast/pull/4521#issuecomment-1384166361

I have worked on code that added the wait-until-available concept that is in the workspace to our OC 3.x archive service, this could be ported to the asset manager so you could initiate a WF on a GLAICER mediapackage and the asset manager would request it from the workspace. The workspace would initiate the restore and tell the asset manager it's unavailable and to poll again. Repeated until the mediapackage is restored.